README.md
# Exit Intent (Desktop + Mobile)

[![version][version]](http://npm.im/exit-intent)
[![MIT License][MIT License]](http://opensource.org/licenses/MIT)
[![Standard][Standard]](http://standardjs.com)
[![Standard Version][Standard Version]](https://github.com/conventional-changelog/standard-version)
[![Size][Size]](https://unpkg.com/exit-intent)
[![Size gzip][Size gzip]](https://unpkg.com/exit-intent)

Exit intent detection library. INCLUDING support for mobile and desktop.

**DESKTOP-behaviour:**

 => trigger intent if mouse leaves body ('mouseleave'-event)
 => trigger after user has been inactive for `showAfterInactiveSecondsDesktop` seconds

**MOBILE-behaviour:**

=> trigger ONLY after user has been inactive for `showAfterInactiveSecondsMobile` seconds

## Usage

```js
import exitIntent from 'exit-intent-mobile'

// Initialise
const removeExitIntent = exitIntent({
  maxDisplays: 99999,                    // default 99999
  eventThrottle: 100,                    // default 200
  showAfterInactiveSecondsDesktop: 60,   // default 60
  showAfterInactiveSecondsMobile: 40,    // default 40
  showAgainAfterSeconds: 10,             // default 10
  onExitIntent: () => {
    console.log('exit-intent triggered')
  },
  debug: false,
})

// Destroy
removeExitIntent()
```

### Options

`maxDisplays` (default 99999)
maximum number of times to trigger.

`eventThrottle` (default 200)
event throttle in milliseconds.

`onExitIntent` (default no-op function)
function to call when an exit intent has been detected.

`showAfterInactiveSecondsDesktop` (default 60 seconds)
If user does NOT move mouse or scroll for nr-of-seconds, `onExitIntent`-function will be called.
To disable the timer, pass **undefined**.

`showAfterInactiveSecondsMobile` (default 40 seconds)
If user does NOT move mouse or scroll for nr-of-seconds, `onExitIntent`-function will be called.
To disable the timer, pass **undefined**.


`showAgainAfterSeconds` (default 10 seconds)
If exit-intend was trigger, pause nr-of-seconds before showing it again. Good to NOT annoy the user.


### Example: Use in React

```
class ExitIntendComponent extends React.Component {
  componentDidMount() {
    this.exitIntend = exitIntent({
      maxDisplays: 99999,
      eventThrottle: 100,
      showAfterInactiveSecondsDesktop: 60,
      showAfterInactiveSecondsMobile: 30,
      onExitIntent: () => {
        this.setState({ show: true })
      },
      debug: false,
    })
  }
  componentWillUnmount() {
    this.exitIntend()  // IMPORTANT: clear timeouts
  }
}
```

### License

MIT

[version]: https://img.shields.io/npm/v/exit-intent.svg
[MIT License]: https://img.shields.io/npm/l/exit-intent.svg
[Standard]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg
[Standard Version]: https://img.shields.io/badge/release-standard%20version-brightgreen.svg
[Size]: https://badges.herokuapp.com/size/npm/exit-intent
[Size gzip]: https://badges.herokuapp.com/size/npm/exit-intent?gzip=true

originally based on https://github.com/danhayden/exit-intent (and https://github.com/richriscunha/Exitent)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

* added possibility to disable the timer(s) (bubenkoff)


src/exit-intent.js
import throttle from 'lodash/throttle'
import isTouchDevice from 'is-touch-device'
import {addWheelListener, removeWheelListener} from 'wheel'
const isDesktop = !isTouchDevice()
/**
 * References
 *  => original https://www.npmjs.com/package/exit-intent
 *  => some tips https://www.thepolyglotdeveloper.com/2018/11/using-exit-intent-listener-manage-popups-page/
 */
export default function ExitIntent (options = {}) {
  const defaultOptions = {
    debug: false,
    maxDisplays: 99999,
    eventThrottle: 200,
    showAfterInactiveSecondsDesktop: 60,
    showAfterInactiveSecondsMobile: 40,
    showAgainAfterSeconds: 10,
    onExitIntent: () => {}
  }
  const config = {...defaultOptions, ...options}
  const log = (...args) => {
    if (config.debug) {
      console.log('[exit-intent-mobile]', ...args)
    }
  }
  // ===========================
  // TRIGGER INTEND
  // ... DISPLAY (only maxDisplays-times)
  let displays = 0
  const doDisplay = () => {
    if (displays < config.maxDisplays) {
      displays++
      log('display onExitIntent', displays)
      config.onExitIntent()
      if (displays >= config.maxDisplays) {
        removeEvents()
      }
    }
  }
  // ... LIMIT display to `config.showAgainAfterSeconds` to make sure that we do NOT bug the user too often
  const display = throttle(doDisplay, config.showAgainAfterSeconds * 1000, {
    trailing: false
  })
  // ===========================
  // EVENT LISTENERS
  // DESKTOP: MOUSEOUT event
  const onMouse = () => {
    if (!(e instanceof MouseEvent)) {
      return;
    }
    log('mouseleave')
    display()
  }
  let onMouseLeaveListener
  if (isDesktop) {
    log('register mouseleave for desktop')
    onMouseLeaveListener = document.body.addEventListener(
      'mouseleave',
      throttle(onMouse, config.eventThrottle),
      false
    )
  }
  // ===========================
  // TIMEOUT (show exit-intend AFTER timeout)
  const timeoutOnDevice = isDesktop
    ? config.showAfterInactiveSecondsDesktop
    : config.showAfterInactiveSecondsMobile
  log('timeoutOnDevice', timeoutOnDevice)
  let timer
  const restartTimer = () => {
    if (typeof timeoutOnDevice === 'undefined') {
      log('display disabled')
      return
    }
    if (timer !== undefined) {
      log('clearTimeout()')
      window.clearTimeout(timer)
    }
    timer = window.setTimeout(() => {
      log('display after timeout')
      display()
    }, timeoutOnDevice * 1000)
  }
  restartTimer() // start initial timer
  // ===========================
  // LISTENERS FOR `restartTimer()`
  const listeners = [] // array to store listeners
  const registerEvent = (event, target) => {
    log('registering event for restartTimer', event, target)
    const listener = target.addEventListener(
      event,
      throttle(event => {
        log('throttled listener', event)
        restartTimer()
      }, config.eventThrottle),
      false
    )
    listeners.push({event, listener, target})
    return listener
  }
  if (isDesktop) {
    registerEvent('scroll', window)
    registerEvent('mousemove', window)
    addWheelListener(window, event => {
      throttle(event => {
        log('throttled wheel listener', event)
        restartTimer()
      }, config.eventThrottle)
    })
  }
  if (isTouchDevice) {
    registerEvent('touchstart', document.body)
    registerEvent('touchend', document.body)
    registerEvent('touchmove', document.body)
  }
  // ===========================
  // CLEANUP
  const removeEvents = () => {
    log('removeEvents', displays)
    if (onMouseLeaveListener) {
      document.body.removeEventListener('mouseleave', onMouseLeaveListener)
      removeWheelListener(window, () => {
        log('removeWheelListener')
      })
    }
    listeners.forEach(theListener => {
      const {event, listener, target} = theListener
      target.removeEventListener(event, listener)
    })
  }
  return removeEvents
}
